### DISCUSSION TITLE
How to set a base url for useFetch globally once

### DISCUSSION THREAD

**Discussion by nlvedwin** [nuxt/nuxt#17531] [Questions] [answered]
so i wanted to use useFetch for my laravel API and i wanted to set the base url to the laravel’s url how i can set the base url once? just like in nuxt2 for axios module

**Comment by zhoutengshen** [comment]
$fetch.create

**Reply by frbuceta** [reply]
I like this definition better but it doesn't work for me in a composable.

```ts
export default $fetch.create({
  baseURL: 'https://xxxxxxxxx.run.app',
  headers: {
    // eslint-disable-next-line quote-props
    // 'Authorization': `Bearer ${token.value}`
  }
})
```

`$fetch is not defined`

**Comment by nlvedwin** [comment]
how about the useFetch?

**Comment by danielroe** [comment, marked as answer]
In that case you can create a wrapper composable:

```ts
import type { UseFetchOptions } from '#app'
const useMyFetch = (url: string, options: UseFetchOptions<unknown>) => useFetch(url, { baseURL: 'https://my.api/', ...options })
```

**Reply by nlvedwin** [reply]
i see, but how would i retain the suggestions/autocomplete for parameters?

**Reply by danielroe** [reply]
Updated.

**Reply by nlvedwin** [reply]
wow thank you dan!👏👏

in the future it would be awesome if there’s an option in the nuxt config to set the api base url that would be cool! 

another option would be adding proxy

**Reply by danielroe** [reply]
There is an option - `app.baseURL` - but note that it changes the entire baseURL of your app, which may not be what you mean. (And `$fetch`/`useFetch` need to stay in sync with your app so you can't set this option to your laravel api.)

**Reply by lucassimines** [reply]
@danielroe how to keep the typing on wrapper function?

I'm trying the following:

`useApiFetch<Page[]>('/pages')`

But the wrapper function doesn't recognize my response type `Page[]`

The error:

```
(alias) interface Page
import Page
Expected 0 type arguments, but got 1.ts(2558)
```

**Reply by peterkimzz** [reply]
Hi @danielroe, your solution works for me. Thanks!
But, I have a quick question. When I use the wrapped composable in server-side hook, nuxt3 said

```
[Vue warn]: onServerPrefetch is called when there is no active component instance to be associated with. 
Lifecycle injection APIs can only be used during execution of setup(). 
If you are using async setup(), make sure to register lifecycle hooks before the first await statement.
```

Can I ignore the warning?

**Comment by Kingside88** [comment]
.env file:
`{
    API_BASE_URL: https://localhost:7257/
}`

composables/useGetFetch.ts:
`export const useGetFetch = (url, opts = {}) => {
  const config = useRuntimeConfig();
  return useFetch(url, {
      ...opts,
      method: "GET",
      baseURL: config.API_BASE_URL,
      credentials: 'omit'
  })
}`
pages/mypage.vue:
`<script setup lang="ts">
let personnels = ref();

const nuxtApp = useNuxtApp()
nuxtApp.hook('page:finish', async () => {
  const { data: personnelsData } = await useGetFetch("personnelsAll");
  personnels.value = personnelsData.value;
})
</script>`

This works greak.
